unit tests for branding admin api

sadatmalik opened the pull request:
- BrandingInfo - unit tests and refactor to use @Builder annotation
- BrandingAdminApi - unit tests and refactor away some boilerplate code
- BrandingServiceImpl - unit tests, some java-doc and minor method refactoring

sadatmalik committed:
adding builder annotation to BrandingInfo with unit test

sadatmalik committed:
unit test for BrandingAdminApi and refactoring some boilerplate code

sadatmalik committed:
unit test for BrandingServiceImpl + some javadoc + minor code refactoring

camerojo commented on server/src/main/java/org/tbbtalent/server/api/admin/BrandingAdminApi.java:
```diff
@@ -16,48 +16,34 @@
      * @return branding information
      */
     @GetMapping()
-    public Map<String, Object> getBrandingInfo() {
```
This method of returning  results from API calls is all through the code and would be a big job to replace. 
We should talk about the pros and cons.

The advantage of the Map<String, Object> and DtoBuilder approach is that we can define subsets of server objects that are passed back to the browser. Whether or not this was a good design approach - it is what we currently have and we would need to think seriously about changing it. There are definitely places in the code which depend on the flexibility of DtoBuilder to work.

sadatmalik replied:
Looking at e.g. the candidate api i see that this would be quite extensive work to replace throughout the code base. Ok let's talk it through. I think advantages might be better type checking, lighter code base, and using Jackson's natural support.

camerojo replied:
Fair enough. One of the trickier ways we use DTO's is when we need to convert a collection of Candidates where the conversion of each candidate in the collection is different based on the user requesting the data. Partners get to see full details of candidates assigned to them, but only partial details of candidates who are not.

sadatmalik replied:
Reverting these changes - will re-introduce in a separate PR as agreed in refinement.

camerojo commented on server/src/test/java/org/tbbtalent/server/api/admin/BrandingAdminApiTest.java:
```diff
@@ -0,0 +1,86 @@
+ */
+@WebMvcTest(BrandingAdminApi.class)
+@AutoConfigureMockMvc
+class BrandingAdminApiTest extends ApiTestBase {
```
Where does ApiTestBase come from and what does it do? I can't find it in any import.

sadatmalik replied:
ApiTestBase is in the same package as BrandingAdminApiTest (it was merged with the AuthAdminApi unit tests). It's just a container for some beans we need to mock out in all admin Api unit test classes.

camerojo replied:
Mmmm - for some reason my Intellij could not find the source. It would be good to track down why.

sadatmalik replied:
That's odd. The PR with ApiTestBase is merged into staging (not yet master). Should get picked up if you pull from staging?

camerojo commented on server/src/main/java/org/tbbtalent/server/model/db/BrandingInfo.java:
```diff
@@ -29,6 +27,7 @@
 @Getter
 @Setter
 @ToString
+@Builder
```
Lets chat about the pros and cons of the Builder pattern.

For example, https://softwareengineering.stackexchange.com/questions/419403/is-it-a-good-or-a-bad-idea-to-use-the-builder-pattern-everywhere

Also is it worth the effort moving everything across to it - given that the existing code doesn't use the pattern extensively.

sadatmalik replied:
Sure let's chat. I brought it in here primarily for use in the unit test. I think it's a neat pattern to introduce where we can easily do so while we are going through writing unit tests. The Lombok annotation makes it very easy but I wouldn't say it's worth the effort using it everywhere.

camerojo commented on server/src/main/java/org/tbbtalent/server/service/db/impl/BrandingServiceImpl.java:
```diff
@@ -34,47 +36,51 @@
     public BrandingInfo getBrandingInfo(@Nullable String partnerAbbreviation) {
 
         User user = userService.getLoggedInUser();
 
```
Why is this better? I think it is a matter of personal preference. We lost some comments in the change here - highlighting for example that the different processing related to whether or not a user is logged in.

sadatmalik replied:
I lifted the comments out as JavaDoc - so don't think they're lost? A learned habit I guess through working with SonarQube which flags this up.
I wouldn't say the code change is better but maybe it is cleaner? That's what i was aiming at anyway! You're right though it's personal preference - this one's another habit learned from SonarQube which encourages the ternary if/else operator where possible.

camerojo replied:
Note that this call to getBrandingInfo is for admin portal - and admin portal uses default branding - ie is not partner specific - note the null passed to the getBrandingInfo call.

Partner based branding is only supplied to the candidate portal. See the implementation of getBrandingInfo when a non null partner is supplied. In particular, partner based branding includes support for different landing pages.

That is why - for the admin portal - the DTO of brandingInfo does not pass back the registrationLandingPage. There is no concept of a landing page for admin portal logins.

I think that this is what was causing you some confusion in earlier discussions. Sorry I did not pick this up before.

sadatmalik replied:
Reverting these changes in this PR - may re-introduce separately if agreed in refinement.

camerojo commented on server/src/main/java/org/tbbtalent/server/api/admin/BrandingAdminApi.java:
```diff
@@ -16,48 +16,34 @@
+        return ResponseEntity.ok(info);
     }
 
-    DtoBuilder brandingInfoDto() {
```
Note that this call to getBrandingInfo is for admin portal - and admin portal using default branding - ie is not partner specific - note the null passed to the getBrandingInfo call.

Partner based branding is only supplied to the candidate portal. See the implementation of getBrandingInfo when a non null partner is supplied. In particular, partner based branding includes support for different landing pages.

That is why - for the admin portal - the DTO of brandingInfo does not pass back the registrationLandingPage. There is no concept of a landing page for admin portal logins.

I think that this is what was causing you some confusion in earlier discussions. Sorry I did not pick this up before.

sadatmalik committed:
reverting builder refactoring - will introduce separately for team review

sadatmalik committed:
updating unit tests - all passing

sadatmalik committed:
reverting imports

camerojo commented on server/src/main/java/org/tbbtalent/server/service/db/impl/BrandingServiceImpl.java:
```diff
@@ -34,15 +36,17 @@
-    }
-
+    /**
+     * Returns the branding information for a partner.
```
We don't duplicate JavaDoc on method implementations that is already present on the method as described in the interface. That leads to duplication which easily gets out of sync.

I would remove these JavaDoc comments

sadatmalik replied:
Yep - makes sense - removed.

camerojo commented on server/src/main/java/org/tbbtalent/server/service/db/impl/BrandingServiceImpl.java:
```diff
@@ -70,8 +74,8 @@ public BrandingInfo getBrandingInfo(@Nullable String partnerAbbreviation) {
-        if (partner instanceof SourcePartner) {
-            info.setLandingPage(((SourcePartner)partner).getRegistrationLandingPage());
+        if (partner instanceof SourcePartner sourcePartner) {
+            info.setLandingPage((sourcePartner).getRegistrationLandingPage());
```
Nice! I didn't know this syntax.

But do you need the (sourcePartner). This would work, right?

```suggestion
            info.setLandingPage(sourcePartner.getRegistrationLandingPage());
```

sadatmalik committed:
Update server/src/main/java/org/tbbtalent/server/service/db/impl/BrandingServiceImpl.java

Ah - these brackets crept back in while removing the builder related refactoring. They're not needed.

Co-authored-by: John Cameron <john@cameronfoundation.org>

sadatmalik committed:
removed duplicated java doc

sadatmalik committed:
Merge branch '19-write-unit-tests-for-branding-admin-api' of https://github.com/talentbeyondboundaries/tbbtalentv2 into 19-write-unit-tests-for-branding-admin-api
Resolve conflicts
